File: cdm-sources/src/cdm/margin/schedule/functions/BuildStandardizedSchedule.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.event.common.Trade import Trade
from cdm.margin.schedule.StandardizedSchedule import StandardizedSchedule
from cdm.margin.schedule.functions.StandardizedScheduleNotional import StandardizedScheduleNotional
from cdm.margin.schedule.functions.StandardizedScheduleAssetClass import StandardizedScheduleAssetClass
from cdm.margin.schedule.functions.StandardizedScheduleProductClass import StandardizedScheduleProductClass
from cdm.margin.schedule.functions.StandardizedScheduleDuration import StandardizedScheduleDuration
from cdm.margin.schedule.functions.StandardizedScheduleNotionalCurrency import StandardizedScheduleNotionalCurrency

__all__ = ['BuildStandardizedSchedule']


@replaceable
def BuildStandardizedSchedule(trade: Trade) -> StandardizedSchedule:
    """
    Takes a trade and uses qualification to extract the relevant information to populate the grid that will be used to calculate the gross initial margin.
    
    Parameters 
    ----------
    trade : Trade
    
    Returns
    -------
    standardizedSchedule : StandardizedSchedule
    
    """
    self = inspect.currentframe()
    
    
    assetClass = StandardizedScheduleAssetClass(rosetta_resolve_attr(self, "trade"))
    productClass = StandardizedScheduleProductClass(rosetta_resolve_attr(self, "trade"))
    standardizedSchedule = _get_rosetta_object('StandardizedSchedule', 'assetClass', rosetta_resolve_attr(self, "assetClass"))
    standardizedSchedule = set_rosetta_attr(rosetta_resolve_attr(self, 'standardizedSchedule'), 'productClass', rosetta_resolve_attr(self, "productClass"))
    standardizedSchedule = set_rosetta_attr(rosetta_resolve_attr(self, 'standardizedSchedule'), 'notional', StandardizedScheduleNotional(rosetta_resolve_attr(self, "trade"), rosetta_resolve_attr(self, "assetClass"), rosetta_resolve_attr(self, "productClass")))
    standardizedSchedule = set_rosetta_attr(rosetta_resolve_attr(self, 'standardizedSchedule'), 'notionalCurrency', StandardizedScheduleNotionalCurrency(rosetta_resolve_attr(self, "trade"), rosetta_resolve_attr(self, "assetClass"), rosetta_resolve_attr(self, "productClass")))
    standardizedSchedule = set_rosetta_attr(rosetta_resolve_attr(self, 'standardizedSchedule'), 'durationInYears', StandardizedScheduleDuration(rosetta_resolve_attr(self, "trade"), rosetta_resolve_attr(self, "assetClass"), rosetta_resolve_attr(self, "productClass")))
    
    
    return standardizedSchedule

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/margin/schedule/functions/BuildStandardizedSchedule.java
package cdm.margin.schedule.functions;

import cdm.event.common.Trade;
import cdm.margin.schedule.StandardizedSchedule;
import cdm.margin.schedule.StandardizedSchedule.StandardizedScheduleBuilder;
import cdm.margin.schedule.StandardizedScheduleAssetClassEnum;
import cdm.margin.schedule.StandardizedScheduleProductClassEnum;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.functions.ModelObjectValidator;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import java.util.Optional;
import javax.inject.Inject;


@ImplementedBy(BuildStandardizedSchedule.BuildStandardizedScheduleDefault.class)
public abstract class BuildStandardizedSchedule implements RosettaFunction {
	
	@Inject protected ModelObjectValidator objectValidator;
	
	// RosettaFunction dependencies
	//
	@Inject protected StandardizedScheduleAssetClass standardizedScheduleAssetClass;
	@Inject protected StandardizedScheduleDuration standardizedScheduleDuration;
	@Inject protected StandardizedScheduleNotional standardizedScheduleNotional;
	@Inject protected StandardizedScheduleNotionalCurrency standardizedScheduleNotionalCurrency;
	@Inject protected StandardizedScheduleProductClass standardizedScheduleProductClass;

	/**
	* @param trade 
	* @return standardizedSchedule 
	*/
	public StandardizedSchedule evaluate(Trade trade) {
		StandardizedSchedule.StandardizedScheduleBuilder standardizedScheduleBuilder = doEvaluate(trade);
		
		final StandardizedSchedule standardizedSchedule;
		if (standardizedScheduleBuilder == null) {
			standardizedSchedule = null;
		} else {
			standardizedSchedule = standardizedScheduleBuilder.build();
			objectValidator.validate(StandardizedSchedule.class, standardizedSchedule);
		}
		
		return standardizedSchedule;
	}

	protected abstract StandardizedSchedule.StandardizedScheduleBuilder doEvaluate(Trade trade);

	protected abstract MapperS<StandardizedScheduleAssetClassEnum> assetClass(Trade trade);

	protected abstract MapperS<StandardizedScheduleProductClassEnum> productClass(Trade trade);

	public static class BuildStandardizedScheduleDefault extends BuildStandardizedSchedule {
		@Override
		protected StandardizedSchedule.StandardizedScheduleBuilder doEvaluate(Trade trade) {
			StandardizedSchedule.StandardizedScheduleBuilder standardizedSchedule = StandardizedSchedule.builder();
			return assignOutput(standardizedSchedule, trade);
		}
		
		protected StandardizedSchedule.StandardizedScheduleBuilder assignOutput(StandardizedSchedule.StandardizedScheduleBuilder standardizedSchedule, Trade trade) {
			standardizedSchedule
				.setAssetClass(assetClass(trade).get());
			
			standardizedSchedule
				.setProductClass(productClass(trade).get());
			
			standardizedSchedule
				.setNotional(standardizedScheduleNotional.evaluate(trade, assetClass(trade).get(), productClass(trade).get()));
			
			standardizedSchedule
				.setNotionalCurrency(standardizedScheduleNotionalCurrency.evaluate(trade, assetClass(trade).get(), productClass(trade).get()));
			
			standardizedSchedule
				.setDurationInYears(standardizedScheduleDuration.evaluate(trade, assetClass(trade).get(), productClass(trade).get()));
			
			return Optional.ofNullable(standardizedSchedule)
				.map(o -> o.prune())
				.orElse(null);
		}
		
		@Override
		protected MapperS<StandardizedScheduleAssetClassEnum> assetClass(Trade trade) {
			return MapperS.of(standardizedScheduleAssetClass.evaluate(trade));
		}
		
		@Override
		protected MapperS<StandardizedScheduleProductClassEnum> productClass(Trade trade) {
			return MapperS.of(standardizedScheduleProductClass.evaluate(trade));
		}
	}
}
